src/compiler/checker.ts

```diff
@@ -33681,9 +33681,33 @@ namespace ts {
                     return grammarErrorOnNode(withMinus ? node.parent : node, diagnosticMessage, literal);
                 }
             }
+
+            // Realism (size) checking
+            checkNumericLiteralValueSize(node);
+
             return false;
         }
 
+        function checkNumericLiteralValueSize(node: NumericLiteral) {
+            // Scientific notation (e.g. 2e54 and 1e00000000010) can't be converted to bigint
+            // Literals with 15 or fewer characters aren't long enough to reach past 2^53 - 1
+            // Fractional numbers (e.g. 9000000000000000.001) are inherently imprecise anyway
+            if (node.numericLiteralFlags & TokenFlags.Scientific || node.text.length <= 15 || node.text.indexOf(".") !== -1) {
+                return;
+            }
+
+            // We can't rely on the runtime to accurately store and compare extremely large numeric values
+            // Even for internal use, we use getTextOfNode: https://github.com/microsoft/TypeScript/issues/33298
+            // Thus, if the runtime claims a too-large number is lower than Number.MAX_SAFE_INTEGER,
+            // it's likely addition operations on it will fail too
+            const apparentValue = +getTextOfNode(node);

```

Comment:
Is this prefix `+` being used to do 0 + "034434543" ? If so, I feel like it might be worth adding the zero (feels like relying on undefined behavior... Open to arguments though.

Reply:
> relying on undefined behavior

Interesting. There are a few places in the codebase that already use this behavior. Searching within `src/**/*.ts.` excluding `.dt.s` files:

* `(= |\()\+` _(regular expression)_ gives 36 results
* `parseInt(` gives 26 results
* `parseFloat(` gives none
* `Number(` _(case sensitive, full word)_ gives 7 results
* `0 + ` and `0+` give nothing used like this

...so I think this actually is more aligned with the existing practices?

Reply:
Yeah, that does sound like it to me. Internally consistent is consistent. Thanks for taking the time.

Let's do it then.